drivers/sensor/sensor_shell.c

```diff
@@ -49,147 +54,136 @@ const char *sensor_channel_name[SENSOR_CHAN_ALL] = {
-			if (!sensor_channel_get(dev, i, value)) {
-				if (i != SENSOR_CHAN_ACCEL_XYZ &&
-				    i != SENSOR_CHAN_GYRO_XYZ &&
-				    i != SENSOR_CHAN_MAGN_XYZ) {
-					shell_print(shell,
-					    "channel idx=%d %s = %10.6f", i,
-					    sensor_channel_name[i],
-					    sensor_value_to_double(&value[0]));
-				} else {
-					shell_print(shell,
-					    "channel idx=%d %s = %10.6f", i,
-					    sensor_channel_name[i],
-					    sensor_value_to_double(&value[0]));
-					shell_print(shell,
-					    "channel idx=%d %s = %10.6f", i,
-					    sensor_channel_name[i],
-					    sensor_value_to_double(&value[1]));
-					shell_print(shell,
-					    "channel idx=%d %s = %10.6f", i,
-					    sensor_channel_name[i],
-					    sensor_value_to_double(&value[2]));
-				}
+	} else {
+		for (int i = 2; i < argc; i++) {
+			err = handle_channel_by_name(shell, dev, argv[i]);
+			if (err < 0) {
+				shell_error(shell,
+					"Failed to read channel (%s)", argv[i]);
 			}
 		}
-
-	} else {
-		shell_error(shell, "Could not bind the %s sensor", argv[1]);
-		return -ENXIO;
 	}
 
 	return 0;
 }
 
-static int cmd_list_sensors(const struct shell *shell,
-			      size_t argc, char **argv)
+static void channel_name_get(size_t idx, struct shell_static_entry *entry);
+
+SHELL_DYNAMIC_CMD_CREATE(dsub_channel_name, channel_name_get);
+
+static void channel_name_get(size_t idx, struct shell_static_entry *entry)
 {
+	int cnt = 0;
+
+	entry->syntax = NULL;
+	entry->handler = NULL;
+	entry->help  = NULL;
+	entry->subcmd = &dsub_channel_name;
+
+	for (int i = 0; i < SENSOR_CHAN_ALL; i++) {
+		if (sensor_channel_name[i] != NULL) {
+			if (cnt == idx) {
+				entry->syntax = sensor_channel_name[i];
+				break;
+			}
+			cnt++;
+		}
+	}
+}
+
+static void device_name_get(size_t idx, struct shell_static_entry *entry);
+
+SHELL_DYNAMIC_CMD_CREATE(dsub_device_name, device_name_get);
+
+static void device_name_get(size_t idx, struct shell_static_entry *entry)
+{
+	int device_idx = 0;
 	struct device *dev;
 
-	ARG_UNUSED(argc);
-	ARG_UNUSED(argv);
+	entry->syntax = NULL;
+	entry->handler = NULL;
+	entry->help  = NULL;
+	entry->subcmd = &dsub_channel_name;
 
-	shell_print(shell, "devices:");
 	for (dev = __device_init_start; dev != __device_init_end; dev++) {
-		if (dev->driver_api != NULL) {
-			shell_print(shell, "- %s", dev->config->name);
+		if ((dev->driver_api != NULL) &&
+		strcmp(dev->config->name, "") && (dev->config->name != NULL)) {
+			if (idx == device_idx) {
+				entry->syntax = dev->config->name;
+				break;
+			}
+			device_idx++;
 		}
 	}
-
-	return 0;
 }
 
 SHELL_STATIC_SUBCMD_SET_CREATE(sub_sensor,
-	SHELL_CMD_ARG(get, NULL, "<device_name> [chanel_idx]", cmd_get_sensor,
-		      2, 1),
-	SHELL_CMD(list, NULL, "List configured sensors", cmd_list_sensors),
-	SHELL_CMD_ARG(list_channels, NULL, "<device_name>",
-		  cmd_list_sensor_channels, 2, 0),
+	SHELL_CMD_ARG(get, &dsub_device_name, SENSOR_GET_HELP, cmd_get_sensor,
```

Comment:
With this change the user needs to know what are the channels supported by a device, auto-completion just lists everything and then it is a trial and error trying to figure out what channels are supported. can we leave the list_channels command to explicitly list channels supported by a sensor?

Reply:
`list_channels` was attempting to read all channels in the device and relied on sensor returning error when channel is invalid. Same functionality can now be achieved using `sensor get <device>` which reads all channels and prints results only for those which did not return error.

Reply:
ah ok, I see this now, thanks.